drivers/display/ek79652.c

```diff
@@ -0,0 +1,552 @@
+#define EK79652_RESET_PIN		DT_INST_GPIO_PIN(0, reset_gpios)
+#define EK79652_RESET_CNTRL		DT_INST_GPIO_LABEL(0, reset_gpios)
+#define EK79652_RESET_FLAGS		DT_INST_GPIO_FLAGS(0, reset_gpios)
+
+#define EPD_PANEL_WIDTH			DT_INST_PROP(0, width)
+#define EPD_PANEL_HEIGHT		DT_INST_PROP(0, height)
+#define EK79652_PIXELS_PER_BYTE		8U
+
+/* Horizontally aligned page! */
+#define EK79652_NUMOF_PAGES		(EPD_PANEL_WIDTH / \
+					 EK79652_PIXELS_PER_BYTE)
+#define EK79652_BUFFER_SIZE		(EK79652_NUMOF_PAGES * EPD_PANEL_HEIGHT)
+#define EK79652_PANEL_FIRST_GATE	0U
+#define EK79652_PANEL_LAST_GATE		(EPD_PANEL_HEIGHT - 1)
+#define EK79652_PANEL_FIRST_PAGE	0U
+#define EK79652_PANEL_LAST_PAGE		(EK79652_NUMOF_PAGES - 1)
+
+struct ek79652_data {
+	const struct ek79652_config *config;
+	const struct device *reset;
+	const struct device *dc;
+	const struct device *busy;
+	bool initial;
+	bool full_refresh;
+	bool disp_refresh;
+};
+
+struct ek79652_config {
+	struct spi_dt_spec bus;
+};
+
+static uint8_t ek79652_softstart[] = DT_INST_PROP(0, softstart);
+static uint8_t ek79652_pwr[] = DT_INST_PROP(0, pwr);
+#if DT_INST_NODE_HAS_PROP(0, pwropt)
+static uint8_t ek79652_pwropt[] = DT_INST_PROP(0, pwropt);
+#endif
+
+static int ek79652_write_data(struct ek79652_data *driver,
+				   uint8_t cmd, uint8_t *data, size_t len,
+				   uint8_t *disp_data, size_t disp_len)
+{
+	struct spi_buf buf = {.buf = &cmd, .len = sizeof(cmd)};
+	struct spi_buf_set buf_set = {.buffers = &buf, .count = 1};
+
+	gpio_pin_set(driver->dc, EK79652_DC_PIN, 1);
+	if (spi_write_dt(&driver->config->bus, &buf_set)) {
+		return -EIO;
+	}
+
+	if (data != NULL) {
+		for (int i = 0 ; i < len ; i++) {
+			buf.buf = &data[i];
+			buf.len = 1;
+			gpio_pin_set(driver->dc, EK79652_DC_PIN, 0);
+			if (spi_write_dt(&driver->config->bus, &buf_set)) {
+				return -EIO;
+			}
+		}
+	}
+
+	if (disp_data != NULL) {
+		for (int i = 0 ; i < disp_len ; i++) {
+			buf.buf = &disp_data[i];
+			buf.len = 1;
+			gpio_pin_set(driver->dc, EK79652_DC_PIN, 0);
+			if (spi_write_dt(&driver->config->bus, &buf_set)) {
+				return -EIO;
+			}
+		}
+	}
+
+	return 0;
+}
+
+static int ek79652_write_cmd(struct ek79652_data *driver,
+				   uint8_t cmd, uint8_t *data, size_t len)
+{
+	return ek79652_write_data(driver, cmd, data, len, NULL, 0);
+}
+
+static void ek79652_busy_wait(struct ek79652_data *driver)
+{
+	int pin = gpio_pin_get(driver->busy, EK79652_BUSY_PIN);
+
+	while (pin > 0) {
+		__ASSERT(pin >= 0, "Failed to get pin level");
+		LOG_DBG("wait %u", pin);
+		k_sleep(K_MSEC(EK79652_BUSY_DELAY));
+		pin = gpio_pin_get(driver->busy, EK79652_BUSY_PIN);
+	}
+}
+
+static int ek79652_update_display(const struct device *dev)
+{
+	struct ek79652_data *driver = dev->data;
+
+	LOG_DBG("Trigger full update sequence");
+
+	if (ek79652_write_cmd(driver, EK79652_CMD_DRF, NULL, 0)) {
+		return -EIO;
+	}
+
+	k_sleep(K_MSEC(EK79652_BUSY_DELAY));
+
+	return 0;
+}
+
+static int set_partial_lut(struct ek79652_data *driver)
```

Comment:
These days, waveform settings should come from controller's OTP, best with consideration of temperature.

Reply:
when I tried with OTP there is heavy ghosting effect happened on the display and also display refresh happened at very slowly. then only opted for this LUT set ,that too is only for partial_display resfresh.